### ISSUE TITLE
API v2 is no longer available for new app registrations

### ISSUE THREAD

**Issue by laf0rge** [peeter123/digikey-api#3]
digikey has closed the path to regitster new clients/applications for API v2.  Instead the have a "contract-breaking" v3 now :/  The change seems to have happened at some point in October 2019.

**Comment by peeter123** [comment]
Thanks for your message, for now the V2 API keeps working for now with already registerd applications. 

However, I'm planning to migrate to the V3 API in the future.

**Comment by JBRORKZ** [comment]
but with little modification in oauth2 i was able to work with new version. 
aslo changed to new url.
but aplication side require much more porting.

self.auth_code=self.auth_code.replace('&scope','')

**Comment by TomOudeNijhuis** [comment]
I have the parts (renamed it to products to match the API call) working. Had to make changes to models, api, client and oauth2. Solved your issue above by changing code to:
```
split_result = urlsplit(self.path)
params = parse_qs(split_result.query)
self.auth_code = params['code'][0]
```
Unfortunately there is no compatibility with the v2 API. Would you like to merge my changes?

**Comment by peeter123** [comment]
@TomOudeNijhuis Thanks, if you can make a pull request on digkey-api-v3 branch I can re-use your modifications!

**Comment by auphofBSF** [comment]
@peeter123 this python api to digikey is great, Thank You and small thanks to @TomOudeNijhuis for the little uoauth2 fixes, I have forked and created a working V3 dev branch forked from digikey-api-v3 https://github.com/auphofBSF/digikey-api/tree/dev20w02apiV3.  The readme example (extended to show search) now works on V3 
It still requires checking and model alignment to Digikey's changes. It also requires refactoring of the term part to product.
Digikey have gone from using the term part to product. So it would probably be good to refactor the term in the model, client and api.   
Before I go ahead and do this, it is worth considering keeping a class part as mirror of product for easy retrofit. Keen to hear from other users on the merit ?
Pending this I would create a pull request

**Comment by peeter123** [comment]
@auphofBSF Thanks for your work, I haven't gotten around to it. I think it is best to indeed keep the part class to avoid breaking applications using the pip module (however, if we switch to V3 they need to create an application at Digikey anyway). It might be nice to implement something like this ([Stack Overflow](https://stackoverflow.com/a/53219607)).

**Comment by auphofBSF** [comment]
> @auphofBSF Thanks for your work, I haven't gotten around to it. I think it is best to indeed keep the part class to avoid breaking applications using the pip module (however, if we switch to V3 they need to create an application at Digikey anyway). It might be nice to implement something like this ([Stack Overflow](https://stackoverflow.com/a/53219607)).

That Idea looks promising, I will try it

**Comment by auphofBSF** [comment]
@peeter123 I have not been able to put the time into supporting V2
My main requirement has been to access the ordersupport api and to this extent I found the digikey api V3 swagger configuration files for each of the API's 
I also new of the swagger code generator that produced an operating python api, so instead of hand coding all the classes, attributes etc I automated the digikey api specification files into customized installed  python api's client packages and provided a wrapper utilizing my forked V3 branch of your digikey_api to handle the OAuth2 protocol. 

digikey spec files ---> swagger_codegen --->  python api classes
https://github.com/auphofBSF/community-digikey-api-v3-lite/blob/DEV/docs/overview-community-digikey-api-v3.svg

I have just finished tidying this up into a one stop installation and have shared it publicly as https://github.com/auphofBSF/community-digikey-api-v3-lite , this automatically install the code generator module https://github.com/auphofBSF/community-digikey-api-codegen-python-clients
It supports the full productinformation and ordersupport of V3 Digikey API. 

Only real local prerequisite is having Java installed as the swagger codegen is a java app and the python package setup calls to first generate the python code.

Installation is simply by `pip  install git+https://github.com/auphofBSF/community-digikey-api-v3-lite@master`   I could look at making this available on PyPi

This now provides a robust V3 API on which a mapping to V2 could be done.  But I don't know how much of  requirement this so would welcome discussion on best way to proceed

Any feedback appreciated

**Comment by lbrolliar** [comment]
@auphofBSF 
I was able to get the community-digikey-api-v3-lite installed with the pip command above. I had to install the community-digikey-api-codegen-python-clients separately.  (This may be due to me not realizing I didn't have Java already installed.)

I think I have a error in my setup at Digikey. The code is installed; however, when I run the code it prints out:
2020-03-10 11:03:04,363 - digikey.oauth2 - WARNING: Token storage does not exist or malformed, creating new.

and the Web browser displays:
Oops 
Looks like something is not right. Please contact your administrator. 
400 - Unknown or invalid client_id 

I am using a Sandbox App at Digikey and I think it may be the Callback URL.  

What do you put in the Callback URL on Digikey?

Thank you in advance for any assistance you might be able to provide.

**Comment by auphofBSF** [comment]
@lbrolliar good to hear you have got it working sofar,  I have never used the digikey sandbox App but I imagine the call back would be similar.  for the call back on the production app I have just used `https://localhost` I remember this was somewhat problematic and it is one of things you do not do repeatedly. 
If you can make any additional notes on the process it would be great to add it to the documentation

**Comment by lbrolliar** [comment]
@auphofBSF thank you for your response.  Hopefully, if I can get past the authentication it will be smooth sailing and help pave the way for others.

I set up a Digikey production app and tested it and the result was the same as with my sandbox app. (e.g. A web page displaying: Oops Looks like something is not right. Please contact your administrator. 400 - Unknown or invalid client_id).

Thinking my client app might not be properly set up,  I opened a Digikey support ticket. 
From my understanding of Digikey's response, the URL to perform the Digikey authentication for V3 API's needs to be of the form given below. (I could not quickly tell where in the code where the authentication URL is set to test this out with the Python code. )  

**Sandbox API URL:**
https://sandbox-api.digikey.com/v1/oauth2/authorize?response_type=code&client_id=123456789abcdefg&redirect_uri=https://client-app-redirect-uri/

**Production API URL:**
https://api.digikey.com/v1/oauth2/authorize?response_type=code&client_id=123456789abcdefg&redirect_uri=https://client-app-redirect-uri/

**Digikey’s support team directed me to:**
https://developer.digikey.com/documentation/oauth

**Note:** 
I was able to get a login/"Request for Approval" screen to come up in a web browser, when I typed the URL of the form above in the browser address bar with my production API Client ID.  (e.g. A URL of the form - https://api.digikey.com/v1/oauth2/authorize?response_type=code&client_id=123456789abcdefg&redirect_uri=https://localhost)

The Digikey Version 2 API URL does not recognize a version 3 client_id.
(e.g. https://sso.digikey.com/as/authorization.oauth2?response_type=code&client_id=123456789abcdefg&redirect_uri=https://localhost)

**Comment by auphofBSF** [comment]
@lbrolliar I am sorry you are experiencing this frustration. I have verified the steps and freshly connected a new instance to a production app. This fresh instance on an Ubuntu Bionic Virtual machine, with a new conda python virtual environment with python 3.6.9 only extra dependencies installed where pandas and tabulate as per the example/example-requirements.txt

1) I confirmed login to http://developer.digikey.com , go to organization and check app is there then view the app details
![image](https://user-images.githubusercontent.com/11743864/76498993-cb821180-64a2-11ea-9252-7e7a850c00b9.png)

2) with the example https://github.com/auphofBSF/community-digikey-api-v3-lite/blob/master/example/example.py
Ensure the DIGIKEY_CLIENT_ID and DIGIKEY_CLIENT_SECRET are set to same as listed in Production APP, also ensure the DIGIKEY_STORAGE_PATH folder exists, it does not get automatically created
![image](https://user-images.githubusercontent.com/11743864/76499543-c7a2bf00-64a3-11ea-9508-5693211bc0c5.png)

3) in a terminal go to the directory where the example.py is saved and run `python example.py`
On my machine Chrome is the default browser and you should be presented with a logon screen
![image](https://user-images.githubusercontent.com/11743864/76499804-4c8dd880-64a4-11ea-8de2-a2a45c973227.png)

4) after successful login , the request for approval should display
![image](https://user-images.githubusercontent.com/11743864/76499951-8d85ed00-64a4-11ea-8b80-50ef4a43d49b.png)

5) after approving you should be alerted to a NET::ERR_CERT_AUTHORITY_INVALID, go into BUTTON (ADVANCED) and agree to proceed to localhost
![image](https://user-images.githubusercontent.com/11743864/76500293-27e63080-64a5-11ea-893c-7bae766ef4de.png)

6) At this point you should get an ERROR:
![image](https://user-images.githubusercontent.com/11743864/76500378-5237ee00-64a5-11ea-93aa-45ae19928829.png)

7) Go back to your terminal where you issued `python example.py`   Ctrl ^C to break the execution that should look like this.
![image](https://user-images.githubusercontent.com/11743864/76500529-90cda880-64a5-11ea-9e88-0a6e39c70f9c.png)

8) rerun `python example.py` no digikey login should appear but the `Request for Approval` should appear again as per 4) , accepting this should give you retrieved Auth code
![image](https://user-images.githubusercontent.com/11743864/76500723-ef932200-64a5-11ea-9f7d-7c86ce5a9769.png)

9) The `example.py` program should have now run through looking something like this
![image](https://user-images.githubusercontent.com/11743864/76501033-70eab480-64a6-11ea-823e-433a465d94de.png)

You will now see the `token_storage.json` has been created in your nominated DIGIKEY_STORAGE_PATH as described in 2).  I am not overly familiar with OAUTH2 and for convenience used the the work that was done in trying to get @peeter123 digikey-api to work with the new DIGIKEY api.  Its my understanding once you have this `token_storage.json` you can then copy it and use in some other location and the token gets automatically refreshed periodically, all this is done with the forked and modified digikey-api automatically installed as a dependency from this branch's last commit https://github.com/auphofBSF/digikey-api/tree/dev20w02apiV3

There needs to be some work done to improve the initial OAUTH2 function of creating the `token_storage.json`, I think the primary issue is getting around the insecure local host  not having a certificate

I did look into it and it was a body of knowledge and integration I could not take on then and probably not now, It would be great to have a contribution to this as it is a fixed and defined function. 

However these steps only are required to setup the initial token and I have just lived with that as there was no apparent issue through on going use.   There may be some issue if an earlier backup of the `token_storage.json` is used.

See how you get on with this detail and hopefully we can Identify at what step you`re failing.

**Comment by lbrolliar** [comment]
Success!!!

@auphofBSF, Thanks for the pics! They helped greatly in debugging.  I noticed the url in your browser was going to auth.digikey.com; however, mine was still going to sso.digikey.com.  Thus, I figured something must be setup incorrectly on my machine.

I did a fresh install into a **new** Python environment everything worked as you detailed above. (Previously, I had installed the community-digikey-api-v3-lite in the same environment where I installed the code for the V2 Digikey API, digikey-api 0.2.1.)  The clean install seems to be the way to go at least for me.

Below details the process I went through get code installed and working on a Windows platform.

Platform: Windows 10, Anaconda Navigator 1.9.12, Spyder IDE 4.0.1.
Steps:
1.	Create new Python 3.6 environment inside of Anaconda Navigator.
2.	Install pandas and tabulate using Anaconda Navigator.
3.	Open the new environment in Anaconda Nav.  Install Anaconda Power Shell and Spyder.
4.	In Anaconda Power Shell install community-digikey-api-v3-lite 
pip install git+https://github.com/auphofBSF/community-digikey-api-v3-lite@master
5.	Create a directory on your drive for storing the token’s
        (e.g. C:\Users\username\digikey-cache-dir)
6.	Launch Spyder.
7.	Modify the three lines in the example script.
8.	Proceed with Step 1 under @auphofBSF’s post above.
Note: I did not have to do the Ctrl-C in the step above.  After I clicked through to the “Error response” page the program exited on its own.
Below is the example display.
![digikey9a](https://user-images.githubusercontent.com/31293021/76546727-014ed800-645a-11ea-95fa-021e708853fa.png)

Thanks again!

**Comment by auphofBSF** [comment]
great news, You're more than welcome,

**Comment by lbrolliar** [comment]
Quick question.  Is there code for doing a keyword search to find the Digikey part number?

**Comment by peeter123** [comment]
Hi All,

I have started work on [digikey-api-v3](/peeter123/digikey-api/tree/digikey-api-v3) based on the swagger autogenerate work done by @auphofBSF (thanks!).

For now only the Product Information API is included and is 'wrapped' so you do not need to think about acces tokens and such.

It should be relatively stable, please try it and if you have any comments or improvements please let me know or submit a pull-request.

**Comment by eeintech** [comment]
@peeter123 I just gave it a shot and, also follwoing @auphofBSF detailed instructions above, and was able to get the V3 to work :)

I know it may seem obvious but in the `digikey-api-v3` branch README, you may want to add the proper way to clone the repo (and also mention that pip won't work right now): `git clone --branch digikey-api-v3 git@github.com:peeter123/digikey-api.git`

Otherwise, the sample code works, I added prints and this is what I'm getting for the first request:
```
2020-06-29 13:03:32,811 - digikey.v3.api -   INFO: Get product details for: 296-6501-1-ND
2020-06-29 13:03:32,811 - digikey.v3.api -  DEBUG: CALL wrapped -> PartSearchApi.product_details_with_http_info
2020-06-29 13:03:34,120 - digikey.v3.api -  DEBUG: Requests remaining: [995/1000]
{'additional_value_fee': 0.0,
 'alternate_packaging': [{'digi_key_part_number': '296-6501-2-ND',
                          'dk_plus_restriction': False,
                          'manufacturer': {'parameter': 'Manufacturer',
                                           'parameter_id': -1,
                                           'value': 'Texas Instruments',
                                           'value_id': '296'},
                          'manufacturer_part_number': 'NE555DR',
                          'manufacturer_public_quantity': 0,
                          'minimum_order_quantity': 2500,
                          'non_stock': False,
                          'packaging': {'parameter': 'Packaging',
                                        'parameter_id': 7,
                                        'value': 'Tape & Reel (TR)',
                                        'value_id': '1'},
                          'product_description': 'IC OSC SGL TIMER 100KHZ '
                                                 '8-SOIC',
                          'quantity_available': 52500,
                          'quantity_on_order': 0,
                          'supplier_direct_ship': False,
                          'unit_price': 0.081},
                         {'digi_key_part_number': '296-6501-6-ND',
                          'dk_plus_restriction': False,
                          'manufacturer': {'parameter': 'Manufacturer',
                                           'parameter_id': -1,
                                           'value': 'Texas Instruments',
                                           'value_id': '296'},
                          'manufacturer_part_number': 'NE555DR',
                          'manufacturer_public_quantity': 0,
                          'minimum_order_quantity': 1,
                          'non_stock': False,
                          'packaging': {'parameter': 'Packaging',
                                        'parameter_id': 7,
                                        'value': 'Digi-Reel®',
                                        'value_id': '243'},
                          'product_description': 'IC OSC SGL TIMER 100KHZ '
                                                 '8-SOIC',
                          'quantity_available': 56150,
                          'quantity_on_order': 0,
                          'supplier_direct_ship': False,
                          'unit_price': 0.36},
                         {'digi_key_part_number': '296-16792-ND',
                          'dk_plus_restriction': False,
                          'manufacturer': {'parameter': 'Manufacturer',
                                           'parameter_id': -1,
                                           'value': 'Texas Instruments',
                                           'value_id': '296'},
                          'manufacturer_part_number': 'NE555D',
                          'manufacturer_public_quantity': 9975,
                          'minimum_order_quantity': 1,
                          'non_stock': False,
                          'packaging': {'parameter': 'Packaging',
                                        'parameter_id': 7,
                                        'value': 'Tube',
                                        'value_id': '6'},
                          'product_description': 'IC OSC SGL TIMER 100KHZ '
                                                 '8-SOIC',
                          'quantity_available': 5332,
                          'quantity_on_order': 0,
                          'supplier_direct_ship': False,
                          'unit_price': 0.41}],
 'associated_products': [],
 'date_last_buy_chance': None,
 'detailed_description': '555 Type, Timer/Oscillator (Single) IC 100kHz 8-SOIC',
 'digi_key_part_number': '296-6501-1-ND',
 'dk_plus_restriction': False,
 'for_use_with_products': [],
 'kit_contents': [],
 'kits': [],
 'lead_status': 'Lead Free',
 'limited_taxonomy': {'children': [{'children': [{'children': [],
                                                  'new_product_count': 0,
                                                  'parameter': 'Categories',
                                                  'parameter_id': -8,
                                                  'product_count': 1,
                                                  'value': 'Clock/Timing - '
                                                           'Programmable '
                                                           'Timers and '
                                                           'Oscillators',
                                                  'value_id': '689'}],
                                    'new_product_count': 0,
                                    'parameter': 'Categories',
                                    'parameter_id': -8,
                                    'product_count': 1,
                                    'value': 'Integrated Circuits (ICs)',
                                    'value_id': '32'}],
                      'new_product_count': 0,
                      'parameter': 'Categories',
                      'parameter_id': -8,
                      'product_count': 1,
                      'value': 'Out of Bounds',
                      'value_id': '0'},
 'manufacturer': {'parameter': 'Manufacturer',
                  'parameter_id': -1,
                  'value': 'Texas Instruments',
                  'value_id': '296'},
 'manufacturer_lead_weeks': '6 weeks',
 'manufacturer_page_url': 'https://www.digikey.com/en/supplier-centers/texas-instruments',
 'manufacturer_part_number': 'NE555DR',
 'manufacturer_public_quantity': 0,
 'mating_products': [],
 'media_links': [{'media_type': 'Datasheets                                                                                                                     ',
                  'small_photo': '',
                  'thumbnail': '',
                  'title': 'NA555, NE555, SA555, SE555',
                  'url': 'http://www.ti.com/general/docs/suppproductinfo.tsp?distId=10&gotoUrl=http%3A%2F%2Fwww.ti.com%2Flit%2Fgpn%2Fne555'},
                 {'media_type': 'Product '
                                'Photos                                                                                                                 ',
                  'small_photo': 'http://media.digikey.com/Renders/Texas%20Instr%20Renders/296;4040047-3;D;8_sml.jpg',
                  'thumbnail': 'http://media.digikey.com/Renders/Texas%20Instr%20Renders/296;4040047-3;D;8_tmb.jpg',
                  'title': '8-SOIC',
                  'url': 'http://media.digikey.com/Renders/Texas%20Instr%20Renders/296;4040047-3;D;8.jpg'},
                 {'media_type': 'PCN Assembly/Origin',
                  'small_photo': '',
                  'thumbnail': '',
                  'title': 'Assembly Site Addition Rev 8/Dec/2015',
                  'url': 'http://media.digikey.com/pdf/PCNs/Texas%20Instruments/PCN20151016001B.pdf'},
                 {'media_type': 'Featured '
                                'Product                                                                                                               ',
                  'small_photo': '',
                  'thumbnail': '',
                  'title': 'xx555 Single Precision Timers',
                  'url': 'https://www.digikey.com/en/product-highlight/t/texas-instruments/xx555-single-precision-timers'},
                 {'media_type': 'PCN Design/Specification',
                  'small_photo': '',
                  'thumbnail': '',
                  'title': 'Mult Devices Font 21/Apr/2018',
                  'url': 'http://media.digikey.com/pdf/PCNs/Texas%20Instruments/PCN20180322004.pdf'},
                 {'media_type': 'PCN Design/Specification',
                  'small_photo': '',
                  'thumbnail': '',
                  'title': 'Mult Devices Font 21/Apr/2018',
                  'url': 'http://media.digikey.com/pdf/PCNs/Texas%20Instruments/PCN20180322004-2.pdf'},
                 {'media_type': 'Design '
                                'Resources                                                                                                               ',
                  'small_photo': '',
                  'thumbnail': '',
                  'title': 'Available In the Digi-Key KiCad Library',
                  'url': 'https://www.digikey.com/en/resources/design-tools/kicad'}],
 'minimum_order_quantity': 1,
 'my_pricing': [],
 'non_stock': False,
 'obsolete': False,
 'packaging': {'parameter': 'Packaging',
               'parameter_id': 7,
               'value': 'Cut Tape (CT)',
               'value_id': '2'},
 'parameters': [{'parameter': 'Packaging',
                 'parameter_id': 7,
                 'value': 'Cut Tape (CT)',
                 'value_id': '2'},
                {'parameter': 'Part Status',
                 'parameter_id': 1989,
                 'value': 'Active',
                 'value_id': '0'},
                {'parameter': 'Type',
                 'parameter_id': 183,
                 'value': '555 Type, Timer/Oscillator (Single)',
                 'value_id': '243489'},
                {'parameter': 'Count',
                 'parameter_id': 435,
                 'value': '-',
                 'value_id': '1'},
                {'parameter': 'Frequency',
                 'parameter_id': 139,
                 'value': '100kHz',
                 'value_id': '68724'},
                {'parameter': 'Voltage - Supply',
                 'parameter_id': 276,
                 'value': '4.5V ~ 16V',
                 'value_id': '208636'},
                {'parameter': 'Current - Supply',
                 'parameter_id': 2095,
                 'value': '10mA',
                 'value_id': 'u10mA'},
                {'parameter': 'Operating Temperature',
                 'parameter_id': 252,
                 'value': '0°C ~ 70°C',
                 'value_id': '38976'},
                {'parameter': 'Package / Case',
                 'parameter_id': 16,
                 'value': '8-SOIC (0.154", 3.90mm Width)',
                 'value_id': '294928'},
                {'parameter': 'Supplier Device Package',
                 'parameter_id': 1291,
                 'value': '8-SOIC',
                 'value_id': '294926'},
                {'parameter': 'Mounting Type',
                 'parameter_id': 69,
                 'value': 'Surface Mount',
                 'value_id': '409393'},
                {'parameter': 'Base Part Number',
                 'parameter_id': 2078,
                 'value': 'NE555',
                 'value_id': '6589'}],
 'primary_datasheet': 'http://www.ti.com/general/docs/suppproductinfo.tsp?distId=10&gotoUrl=http%3A%2F%2Fwww.ti.com%2Flit%2Fgpn%2Fne555',
 'primary_photo': 'https://media.digikey.com/Renders/Texas%20Instr%20Renders/296;4040047-3;D;8.jpg',
 'primary_video': '',
 'product_description': 'IC OSC SGL TIMER 100KHZ 8-SOIC',
 'product_status': 'Active',
 'product_url': 'https://www.digikey.com/product-detail/en/texas-instruments/NE555DR/296-6501-1-ND/372490',
 'quantity_available': 56144,
 'quantity_on_order': 0,
 'ro_hs_status': 'ROHS3 Compliant',
 'rohs_subs': [],
 'search_locale_used': {'currency': 'USD',
                        'language': 'en',
                        'ship_to_country': 'US',
                        'site': 'US'},
 'series': {'parameter': 'Series',
            'parameter_id': -5,
            'value': '-',
            'value_id': '818'},
 'standard_package': 1,
 'standard_pricing': [{'break_quantity': 1,
                       'total_price': 0.36,
                       'unit_price': 0.36},
                      {'break_quantity': 10,
                       'total_price': 2.75,
                       'unit_price': 0.275},
                      {'break_quantity': 25,
                       'total_price': 6.19,
                       'unit_price': 0.2476},
                      {'break_quantity': 100,
                       'total_price': 17.1,
                       'unit_price': 0.171},
                      {'break_quantity': 250,
                       'total_price': 36.0,
                       'unit_price': 0.144},
                      {'break_quantity': 500,
                       'total_price': 58.5,
                       'unit_price': 0.117},
                      {'break_quantity': 1000,
                       'total_price': 90.0,
                       'unit_price': 0.09}],
 'suggested_subs': [],
 'supplier_direct_ship': False,
 'tariff_description': '',
 'unit_price': 0.36}
```

**Comment by kjgoodrick** [comment]
Hi Peter,

I've also been doing a little work based off @auphofBSF's work, though not the swagger version. Primarily this involves adding support for filtering and sorting. You can see that in my fork here: https://github.com/kjgoodrick/digikey-api/tree/dev20w02apiV3 

I looked at trying to add this in off the swagger version, but it wasn't clear to me how that would work with the autogenerated code. I'm not sure if this made obsolete by your new version, but I'll play around with it and see.

**Comment by peeter123** [comment]
@kjgoodrick I think that should be possible in the autogenerated version, you can pass a KeywordSearchRequest model as a kwargs to the  keyword_search function. The KeywordSearchRequest model also has an entry for adding Filters and SortParameters. You can check the available models in [digikey/v3/productinformation/models](/peeter123/digikey-api/tree/digikey-api-v3/digikey/v3/productinformation/models)

**Comment by RainZhangZZFLS** [comment]
Hi Peter, 
Thanks for your contribution to this digiky-api. I have tried [digikey-api-v3](https://github.com/peeter123/digikey-api/tree/digikey-api-v3) and it seems git clone directly does not work, so I just simply download the ZIP and pip install it. Now the problem is the client ID and client Secret provided by the SandBox still does not work. 
![image](https://user-images.githubusercontent.com/36270330/86477897-a19e7480-bd49-11ea-97d2-4627162f9186.png)

I still received the error message like this
`2020-07-03 16:07:49,055 - digikey.oauth.oauth2 - WARNING: Token storage does not exist or malformed, creating new.`
Thus I am commenting to see if any possible solution, I notice there exists some success cases.

**Comment by peeter123** [comment]
@RainZhangZZFLS You created a sandbox application, you need to create a production application for it to work.

You can however now use sandbox API as well (see updated branch)

**Comment by RainZhangZZFLS** [comment]
> @RainZhangZZFLS You created a sandbox application, you need to create a production application for it to work.
> 
> You can however now use sandbox API as well (see updated branch)

That finally works! Thanks Peter! Thanks Anthony for swagger part!

**Comment by RainZhangZZFLS** [comment]
Hi Peter, I have another question on how to extract certain information from the 
`result = digikey.keyword_search(body=search_request)`
For example, like  'manufacturer_part_number': 'CRCW080510K0FKEA', which can be displayed using 'print'. But how to display it separately or output as a string?

**Comment by eeintech** [comment]
@RainZhangZZFLS To save and access data in a dictionary, you'll need to append the `.to_dict()` to the request:
```
result = digikey.keyword_search(body=search_request).to_dict()
```
Then you can use `result['manufacturer_part_number']` to retrieve the part number string.

**Comment by RainZhangZZFLS** [comment]
> @RainZhangZZFLS To save and access data in a dictionary, you'll need to append the `.to_dict()` to the request:
> 
> ```
> result = digikey.keyword_search(body=search_request).to_dict()
> ```
> 
> Then you can use `result['manufacturer_part_number']` to retrieve the part number string.

Thanks for your reply! That really helps!

**Comment by eeintech** [comment]
@peeter123 Are you planning on merging the `digikey-api-v3` branch to master?

**Comment by peeter123** [comment]
@eeintech Yes I think that should be possible on short term.

NOTE: only the partsearch API has been wrapped / implemented

**Comment by peeter123** [comment]
I have released 0.4.0 with support for the PartSearch v3 API, 

If you are interested in helping implementing the other APIs please create a pull request👍
